src/gateway_uri.rs

```diff
@@ -1,46 +1,147 @@
+use std::str::FromStr;
+
+use http::uri::{Authority, Scheme};
 use http::Uri;
 
 use crate::error::BoxError;
+
 /// A normalized gateway origin URI with a default port if none is specified.
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct GatewayUri(Uri);
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct GatewayUri {
+    scheme: Scheme,
+    authority: Authority,
+}
 
 impl GatewayUri {
-    pub fn new(mut gateway_origin: Uri) -> Result<Self, BoxError> {
-        let (scheme, default_port) = match gateway_origin.scheme_str() {
-            Some("http") => ("http", 80),
-            Some("https") | None => ("https", 443),
-            _ => return Err("Unsupported URI scheme".into()),
+    pub fn new(scheme: Scheme, authority: Authority) -> Result<Self, BoxError> {
+        let default_port = if scheme == Scheme::HTTP {
+            80
+        } else if scheme == Scheme::HTTPS {
+            443
+        } else {
+            return Err("Unsupported URI scheme".into());
         };
 
-        if gateway_origin.authority().map(|a| a.port().is_none()).unwrap_or(true) {
-            let authority = if let Some(auth) = gateway_origin.authority() {
-                format!("{}:{}", auth.host(), default_port)
-            } else {
-                return Err("URI must have an authority".into());
-            };
+        // If no explicit port is provided, make the default one explicit
+        let mut authority = authority;
+        if authority.port().is_none() {
+            authority = Authority::from_str(&format!("{}:{}", authority.host(), default_port))
+                .expect("setting default port must succeed");
+        }
 
-            let path_and_query = gateway_origin
-                .path_and_query()
-                .map(|pq| pq.to_string())
-                .unwrap_or_else(|| "/".to_string());
+        Ok(Self { scheme, authority })
+    }
 
-            let builder =
-                Uri::builder().scheme(scheme).authority(authority).path_and_query(path_and_query);
+    // TODO remove
+    pub fn authority(&self) -> &Authority { &self.authority }
 
-            gateway_origin = builder.build()?;
-        }
+    pub fn from_static(string: &'static str) -> Self {
+        Uri::from_static(string).try_into().unwrap()
```

Comment:
Before the unwrap was part of the connect.rs tests. Now that it's part of the public gateway_uri api might might it make more sense to use a documenting `.expect`?